homeassistant/components/pvpc_hourly_pricing/sensor.py

```diff
@@ -0,0 +1,195 @@
+
+    unit_of_measurement = UNIT
+    icon = ICON
+    should_poll = False
+
+    def __init__(self, name, unique_id, pvpc_data_handler):
+        """Initialize the sensor object."""
+        self._name = name
+        self._unique_id = unique_id
+        self._pvpc_data = pvpc_data_handler
+        self._num_retries = 0
+
+        self._init_done = False
+        self._hourly_tracker = None
+        self._price_tracker = None
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Cancel listeners for sensor updates."""
+        self._hourly_tracker()
+        self._price_tracker()
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state:
+            self._pvpc_data.state = state.state
+
+        # Update 'state' value in hour changes
+        self._hourly_tracker = async_track_time_change(
+            self.hass, self.async_update, second=[0], minute=[0]
+        )
+        # Update prices at random time, 2 times/hour (don't want to upset API)
+        random_minute = randint(1, 29)
+        mins_update = [random_minute, random_minute + 30]
+        self._price_tracker = async_track_time_change(
+            self.hass, self.async_update_prices, second=[0], minute=mins_update,
+        )
+        _LOGGER.debug(
+            "Setup of price sensor %s (%s) with tariff '%s', "
+            "updating prices each hour at %s min",
+            self.name,
+            self.entity_id,
+            self._pvpc_data.tariff,
+            mins_update,
+        )
+        await self.async_update_prices(dt_util.utcnow())
+        self._init_done = True
+        await self.async_update_ha_state(True)
+
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID."""
+        return self._unique_id
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._pvpc_data.state
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._pvpc_data.state_available
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._pvpc_data.attributes
+
+    async def async_update(self, *_args):
+        """Update the sensor state."""
+        if not self._init_done:
+            # abort until added_to_hass is finished
+            return
+
+        now = dt_util.utcnow()
+        if self._pvpc_data.process_state_and_attributes(now):
+            await self.async_update_ha_state()
+        else:
+            # If no prices present, download and schedule a future state update
+            self._pvpc_data.state_available = False
+            self.async_schedule_update_ha_state()
```

Comment:
Does this need to call update again or do you already have all the data you need to call `self.async_write_ha_state()`?

Reply:
As we are flagging `unavailable`, we have all the data we need. 
In that case, `self.async_schedule_update_ha_state()` is just calling `self.async_write_ha_state()`, but I see that the latter is preferred, right? I'll change to `self.async_write_ha_state()`.

Reply:
Yes, thats the idea :)